rfcs/20200306-single-client-parameter-server.md

```diff
@@ -0,0 +1,517 @@
+
+
+## Implementation
+
+
+### Low-level Primitives
+
+We can potentially expose them in the future when they are more stable and when we want to allow more advanced use cases.
+
+We will have `Cluster` and `Worker` classes to encapsulate logic related to remote function scheduling.
+
+
+```python
+class Cluster(object):
+
+  def __init__(self, cluster_resolver, failure_handler=None):
+    """Create the cluster instance and connect to the remote cluster."""
+    pass
+
+  @property
+  def workers(self):
+    """Return all available workers."""
+    return self._workers
+
+  def schedule(self, function, args=None, kwargs=None):
+    """Schedule the function on one worker.
+
+    It adds the function to the global scheduling queue and returns future
+    objects immediately.
+    """
+    pass
+
+  def join(self):
+    """Block until all scheduled functions are complete."""
+    pass
+```
+
+
+We will probably merge this `Worker` with executors.
+
+
+```python
+class Worker(object):
+
+  def __init__(self,
+               worker_job_name,
+               cluster,
+               max_scheduled_functions=100):
+    """Create a scheduling queue and a thread that processes the queue."""
+    pass
+
+  def schedule(self, function, args=None, kwargs=None):
+    """Schedule the function on the worker.
+
+    It adds the function to the scheduling queue. It returns Future object
+    immediately unless the scheduling queue is full.
+    """
+    pass
+
+  def healthy(self):
+    """Return a boolean indicating whether the worker is health or not."""
+    pass
+
+  def _set_dead(self):
+    """Declare the worker is dead and poison all future objects."""
+    pass
+
+  def _rebuild_resources(self):
+    """Rebuild worker-local resources when it is recovered from failure."""
+    pass
+```
+
+
+As we mentioned the return value of `schedule` will be `Future` objects if `function` has return values. The `Future` works as a container and will be later-binded with states of either success or complete failure. Overall, this `Future` class has the following benefits:
+
+
+
+*   It allows `schedule` method return immediately after pushing functions to its scheduling queue. It allows these methods to return without needing to wait for acknowledgement from workers.
+*   It serves as the container for values or errors. It would be binded with a value or an error later. When it is rebuilt, we can replace its underlying value silently.
+*   When being passed to `local_result`, we flag it to indicate that this value needs to be fetched eagerly.
+*   (Future work) It captures the lineage between functions and return values so that we can rebuild any poisoned objects. 
+
+```python
+class Future(object):
```

Comment:
So no `on_complete` asynchronous callbacks on futures? The clients check the value of futures exclusively by joining/polling?

Reply:
We can definitely asynchronous callbacks if this is needed. Maybe you can give me some use cases?